Sources_v3/StreamChat/ChatClient.swift

```diff
@@ -271,54 +271,195 @@ public class _ChatClient<ExtraData: ExtraDataTypes> {
-            guard let token = token else {
-                completion(ClientError.MissingToken("Can't connect because `TokenProvider` didn't return a valid token."))
-                return
-            }
-            self?.currentToken = token
-            completion(nil)
-        })
-    }
-    
+}
+
+private extension _ChatClient {
     func createBackgroundWorkers() {
         backgroundWorkers = workerBuilders.map { builder in
             builder(self.databaseContainer, self.apiClient)
         } + eventWorkerBuilders.map { builder in
             builder(self.databaseContainer, self.eventNotificationCenter, self.apiClient)
         }
     }
-    
+
+    func removeConnectionWaitersPassing(_ connectionId: String?) {
+        _connectionIdWaiters.mutate { waiters in
+            waiters.forEach { $0(connectionId) }
+            waiters.removeAll()
+        }
+    }
+
+    func removeTokenWaitersPassing(_ token: Token?) {
+        _tokenWaiters.mutate { waiters in
+            waiters.forEach { $0(token) }
+            waiters.removeAll()
+        }
+    }
+
+    func removeConnectionWaiters() {
+        _connectionIdWaiters.mutate {
+            $0.removeAll()
+        }
+    }
+
+    func removeTokenWaiters() {
+        _tokenWaiters.mutate {
+            $0.removeAll()
+        }
+    }
+
     /// Before CIS-555, `webSocketClient` was initialized in `createBackgroundWorkers` in `init` since it was passed
     /// to the background workers. Now, background workers do not access `webSocketClient` and it's not being
     /// initialized in `init`, so we initialize it here explicitly.
-    private func createWebSocketClient() {
+    func createWebSocketClient() {
         _ = webSocketClient
     }
+
+    func prepareEnvironment(newToken: Token, completion: @escaping (Error?) -> Void) {
+        // Check token is for current user.
+        guard newToken.userId != currentUserId else {
+            // Update the current token with the new one.
+            _currentToken.mutate { $0 = newToken }
+            // Pass new token to waiters.
+            removeTokenWaitersPassing(newToken)
+            // We're done.
+            completion(nil)
+            return
+        }
+
+        // Setting a new user is not possible in connectionless mode.
+        guard config.isClientInActiveMode else {
+            completion(ClientError.ClientIsNotInActiveMode())
+            return
+        }
+
+        // Disconnect from web-socket to stop receiving the events.
+        disconnect()
+
+        // Re-create backgroundWorker's since they are related to the previous user.
+        createBackgroundWorkers()
+
+        // Cancel all API requests since they are related to the previous user.
+        removeTokenWaitersPassing(nil)
+
+        // Update the current token with the new one.
+        _currentToken.mutate { $0 = newToken }
+
+        // Update the current user id to the new one.
+        _currentUserId.mutate { $0 = newToken.userId }
```

Comment:
You don't need to do this if the wrapped type is not a collection. `currentToken = newToken` and `currentUserId = newToken.userId` are equally thread-safe to this.

Reply:
fixed